File: javascript/tests/test_security.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, writeFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';
import { scan } from '../src/security.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURES = join(__dirname, 'fixtures');

test('hardcoded api key in fixture', () => {
  const findings = scan(join(FIXTURES, 'sample.py'), 'Python');
  const secrets = findings.filter(f => f.type === 'hardcoded_secret');
  assert.ok(secrets.length >= 1, `Expected hardcoded_secret, got: ${JSON.stringify(findings)}`);
});

test('eval in js fixture', () => {
  const findings = scan(join(FIXTURES, 'sample.js'), 'JavaScript');
  const evals = findings.filter(f => f.pattern === 'eval_call');
  assert.ok(evals.length >= 1, `Expected eval_call, got: ${JSON.stringify(findings)}`);
});

test('secret severity is error', () => {
  const findings = scan(join(FIXTURES, 'sample.py'), 'Python');
  const secrets = findings.filter(f => f.type === 'hardcoded_secret');
  assert.ok(secrets.every(f => f.severity === 'error'));
});

test('dangerous call severity is warning', () => {
  const findings = scan(join(FIXTURES, 'sample.js'), 'JavaScript');
  const dangerous = findings.filter(f => f.type === 'dangerous_call');
  assert.ok(dangerous.every(f => f.severity === 'warning'));
});

test('findings have required fields', () => {
  const findings = scan(join(FIXTURES, 'sample.py'), 'Python');
  assert.ok(findings.length > 0);
  for (const f of findings) {
    assert.ok('type' in f);
    assert.ok('severity' in f);
    assert.ok('file' in f);
    assert.ok('line' in f);
    assert.ok('pattern' in f);
    assert.ok('detail' in f);
    assert.ok('snippet' in f);
  }
});

test('aws key pattern', () => {
  const dir = mkdtempSync(join(tmpdir(), 'scanner-test-'));
  const path = join(dir, 'config.py');
  writeFileSync(path, 'aws_access_key_id = "AKIAIOSFODNN7EXAMPLE123"\n');
  const findings = scan(path, 'Python');
  const aws = findings.filter(f => f.pattern === 'aws_key');
  assert.ok(aws.length >= 1);
});

test('password pattern', () => {
  const dir = mkdtempSync(join(tmpdir(), 'scanner-test-'));
  const path = join(dir, 'config.py');
  writeFileSync(path, 'password = "mysecretpassword"\n');
  const findings = scan(path, 'Python');
  const pw = findings.filter(f => f.pattern === 'password_assignment');
  assert.ok(pw.length >= 1);
});

File: python/tests/test_security.py
import os
import tempfile
import pytest
from scanner.security import scan

FIXTURES = os.path.join(os.path.dirname(__file__), 'fixtures')


def test_hardcoded_api_key_in_fixture():
    findings = scan(os.path.join(FIXTURES, 'sample.py'), 'Python')
    secrets = [f for f in findings if f['type'] == 'hardcoded_secret']
    assert len(secrets) >= 1, f"Expected hardcoded_secret, got: {findings}"


def test_eval_in_js_fixture():
    findings = scan(os.path.join(FIXTURES, 'sample.js'), 'JavaScript')
    evals = [f for f in findings if f['pattern'] == 'eval_call']
    assert len(evals) >= 1, f"Expected eval_call, got: {findings}"


def test_secret_severity_is_error():
    findings = scan(os.path.join(FIXTURES, 'sample.py'), 'Python')
    secrets = [f for f in findings if f['type'] == 'hardcoded_secret']
    assert all(f['severity'] == 'error' for f in secrets)


def test_dangerous_call_severity_is_warning():
    findings = scan(os.path.join(FIXTURES, 'sample.js'), 'JavaScript')
    dangerous = [f for f in findings if f['type'] == 'dangerous_call']
    assert all(f['severity'] == 'warning' for f in dangerous)


def test_findings_have_required_fields():
    findings = scan(os.path.join(FIXTURES, 'sample.py'), 'Python')
    assert len(findings) > 0
    for f in findings:
        assert 'type' in f
        assert 'severity' in f
        assert 'file' in f
        assert 'line' in f
        assert 'pattern' in f
        assert 'detail' in f
        assert 'snippet' in f


def test_aws_key_pattern():
    content = 'aws_access_key_id = "AKIAIOSFODNN7EXAMPLE123"\n'
    with tempfile.NamedTemporaryFile(suffix='.py', mode='w', delete=False) as f:
        f.write(content)
        path = f.name
    try:
        findings = scan(path, 'Python')
        aws = [x for x in findings if x['pattern'] == 'aws_key']
        assert len(aws) >= 1
    finally:
        os.unlink(path)


def test_password_pattern():
    content = 'password = "mysecretpassword"\n'
    with tempfile.NamedTemporaryFile(suffix='.py', mode='w', delete=False) as f:
        f.write(content)
        path = f.name
    try:
        findings = scan(path, 'Python')
        pw = [x for x in findings if x['pattern'] == 'password_assignment']
        assert len(pw) >= 1
    finally:
        os.unlink(path)


def test_os_system_pattern():
    content = 'import os\nos.system("ls")\n'
    with tempfile.NamedTemporaryFile(suffix='.py', mode='w', delete=False) as f:
        f.write(content)
        path = f.name
    try:
        findings = scan(path, 'Python')
        sys_calls = [x for x in findings if x['pattern'] == 'os_system']
        assert len(sys_calls) >= 1
    finally:
        os.unlink(path)
